components/content-service/pkg/storage/s3.go

```diff
@@ -0,0 +1,417 @@
+		return false, err
+	}
+	return true, nil
+}
+
+// ObjectHash implements PresignedAccess
+func (rs *PresignedS3Storage) ObjectHash(ctx context.Context, bucket string, obj string) (string, error) {
+	resp, err := rs.client.GetObjectAttributes(ctx, &s3.GetObjectAttributesInput{
+		Bucket:           &rs.Config.Bucket,
+		Key:              aws.String(obj),
+		ObjectAttributes: []types.ObjectAttributes{types.ObjectAttributesEtag},
+	})
+	var nsk *types.NoSuchKey
+	if errors.As(err, &nsk) {
+		return "", ErrNotFound
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	return *resp.ETag, nil
+}
+
+// SignDownload implements PresignedAccess
+func (rs *PresignedS3Storage) SignDownload(ctx context.Context, bucket string, obj string, options *SignedURLOptions) (info *DownloadInfo, err error) {
+	resp, err := rs.client.GetObjectAttributes(ctx, &s3.GetObjectAttributesInput{
+		Bucket:           &rs.Config.Bucket,
+		Key:              aws.String(obj),
+		ObjectAttributes: []types.ObjectAttributes{types.ObjectAttributesObjectSize},
+	})
+
+	var nsk *types.NoSuchKey
+	if errors.As(err, &nsk) {
+		return nil, ErrNotFound
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := rs.PresignedFactory().PresignGetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(rs.Config.Bucket),
+		Key:    aws.String(obj),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &DownloadInfo{
+		Meta: ObjectMeta{
+			// TODO(cw): implement this if we need to support FWB with S3
+		},
+		Size: resp.ObjectSize,
+		URL:  req.URL,
+	}, nil
+}
+
+// SignUpload implements PresignedAccess
+func (rs *PresignedS3Storage) SignUpload(ctx context.Context, bucket string, obj string, options *SignedURLOptions) (info *UploadInfo, err error) {
+	resp, err := rs.PresignedFactory().PresignPutObject(ctx, &s3.PutObjectInput{
+		Bucket: &rs.Config.Bucket,
+		Key:    aws.String(obj),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &UploadInfo{
+		URL: resp.URL,
+	}, nil
+}
+
+func newDirectS3Access(client S3Client, config S3Config) *s3Storage {
+	return &s3Storage{
+		Config: config,
+		client: client,
+	}
+}
+
+type s3Storage struct {
+	Config S3Config
+
+	OwnerID, WorkspaceID, InstanceID string
+
+	client S3Client
+}
+
+// Bucket implements DirectAccess
+func (s3st *s3Storage) Bucket(userID string) string {
+	return s3st.Config.Bucket
+}
+
+// BackupObject implements DirectAccess
+func (s3st *s3Storage) BackupObject(name string) string {
+	return s3st.objectName(name)
+}
+
+// Download implements DirectAccess
+func (s3st *s3Storage) Download(ctx context.Context, destination string, name string, mappings []archive.IDMapping) (found bool, err error) {
+	return s3st.download(ctx, destination, s3st.objectName(name), mappings)
+}
+
+// DownloadSnapshot implements DirectAccess
+func (s3st *s3Storage) DownloadSnapshot(ctx context.Context, destination string, name string, mappings []archive.IDMapping) (found bool, err error) {
+	return s3st.download(ctx, destination, name, mappings)
```

Comment:
Do we need to [parse the snapshot](https://github.com/gitpod-io/gitpod/blob/b774d7af3901a13482c5b24aa3ace17d7541d61a/components/content-service/pkg/storage/minio.go#L229) for the download snapshot method before using it?

Reply:
It works without.